rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+            If, instead, bins is a list of floats, then those are
+            the bin boundary values and nothing is to be learned.
+            The width of the output will in that case be the len(bins) + 1.
+        strategy: callable | 'quantiles'
+            If strategy is the string 'quantiles' (default),
+            then bin boundaries will be learned such that each bin
+            receives an approximately equal number of sample input values.
+            ‘Strategy’ may also be a callable that takes
+            (float value, list[float] boundaries) and returns
+            an int bucket_index which represents
+            which bucket to map ‘value’ to.
+        sparse: If True, the layer will output a SparseTensor.
+            Otherwise it will be dense.
+            This does not change the shape or structure of the output.
+            Specifically tf.sparse.to_dense(output) will be the same for both.
+
+    Input shape and type:
+        dtype: floating point.
+        shape: [batch_size, ..., features]
+
+    Output shape and type:
+        dtype: int
+        shape: [batch_size, ..., features, num_bins]
+            i.e., the same as the input shape,
+            with an additional dimension corresponding to
+            the number of bins, which is equal to either
+            the bins constructor argument (if it is an integer),
+            or the length of the bins constructor argument plus 1,
+            if it is a list.
+
+    What happens in `adapt`:
+        We use a streaming quantile estimator to update the bin boundaries
+        so that statistically an element is about equally likely
+        to fall into any bin.
+        Multiple calls to update continue to mutate
+        the layer based on all data seen so far.
+    """
+```
+
+#### TextVectorization
+
+This layer has basic options for managing text in the Keras model.
+It is expected that more advanced users needing custom control will uses Keras-compatible layers provided by tf.text.
+
+Transform a batch of strings (one sample = one string) into either a list of token indices
+(one sample = 1D int tensor or a dense representation (1 sample = 1D float vector).
+
+The processing of each sample unfolds as:
+- Standardize each sample (usually lowercasing + punctuation stripping)
+- Split each sample into substrings (usually words)
+- Recombine substrings into tokens (usually ngrams)
+- Index tokens (associate a unique int value with each token)
+- Transform each sample using this index, either into a vector of ints or a dense float vector.
+
+
+```python
+def __init__(self,
+             tokens=None,
+             standardize='lower_and_strip_punctuation',
+             split='whitespace',
+             ngrams=1,
+             mode='int',
+             max_length=None):
+    """Transforms text into dense vectors or sequences of word indices.
+
+    Arguments:
+        tokens: None (default) | int | list<string>
```

Comment:
You're using "tokens" to mean a list of possible string types.  This isn't how the word "token" has traditionally been used/defined in the NLP field. It was initially introduced to stress the token/type distinction (https://en.wikipedia.org/wiki/Type%E2%80%93token_distinction) when handling text -- e.g., "the dog saw the cat" has 5 word tokens, but 4 word types.  E.g., see https://nlp.stanford.edu/IR-book/html/htmledition/tokenization-1.html or https://www.nltk.org/book/ch01.html#counting-vocabulary.

I therefore think "tokens" should be renamed to "vocabulary" or perhaps "terms."

Reply:
This will be discussed in the design review. My take, in summary, is two-fold:

- The concern with using "vocabulary" or "terms" is that it implicitly refers to words, whereas we mean atoms to be arbitrary substrings (including ngrams, subwords, and even characters). Meanwhile it is well accepted that "tokens" can be arbitrary atomic substrings; subwords 
 are a popular choice, and the literature even refers to ["character-level tokenization"](https://www.google.com/search?q=site%3Aarxiv.org+"character-level+tokenization").

- The distinction between atom instances (tokens) and atom types is not confusing for users, and in NLP software the term "token" seems to be often used to refer to unique atoms (types), perhaps because in either case they are encoded as the exact same Python string (so a string instance like "the cat" is both a token, if found in a document, and a type, if used as a key in a dict, but it's the same string). If you pass a list of atomic strings to a tokenizer constructor, it is clear that these refer to unique types. It's not dissimilar to the fact that e.g. `2` refers simultaneously to to the abstract int "2" type, and to specific instances of a 2.

If you're going to call the process "tokenization" and refer to the atoms that you split your string into as "tokens", then it's perfectly reasonable to pass a "tokens" argument in the constructor of the layer.